<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN"
        "http://www.w3.org/TR/REC-html40/strict.dtd"><html>
<head>
<title>Unison Protein Database</title>
<link rel="shortcut icon" href="av/favicon.png">
<link rel="stylesheet" type="text/css" href="styles/unison.css">
</head>

<body>

<!-- quicklinks start -->
<SCRIPT TYPE="text/javascript" SRC="searchplugin/webinstall.js"></script>
<table width="100%" border=0>
  <tr>
	<td valign="top">
	  <span style="font-size: 150%; font-weight: bold;">Welcome to</span>
	  <br><img src="av/logo5.png" alt="[Unison Logo]">
	  <br><i>integrated, precomputed proteomic predictions for rapid
	  feature-based mining, sequence analysis, and hypothesis generation</i>
	</td>

	<td valign="top" align="right">
	 <table class=quicklinks>
	  <tr><td colspan=2 class="quicklinks_title" align=center><b>quick
		 links</b></td></tr>

	  <tr>
	   <td class="quicklinks_title">info</td>
	   <td class=quicklinks>
	     <a href="index.html">home</a> 
       | <a href="moreinfo.html">more info</a>
  <!-- | <a href="docs/index.html">docs</a> -->
       | <a href="tour/index.html">tour</a> | <a href="http://sourceforge.net/project/screenshots.php?group_id=140591">screenshots</a>
	   | <a href="license.html">license</a> | <a href="credits.html">credits</a>
	   </td>
	  </tr>

	  <tr>
	   <td class="quicklinks_title">contacts</td>
	   <td class=quicklinks><a href="http://sourceforge.net/tracker/?atid=759616&amp;group_id=140591&amp;func=browse">issue tracker</a> | <a href="http://lists.sourceforge.net/lists/listinfo/unison-db-announce">announcment mailing list</a></td>
	  </tr>

	  <tr>
	   <td class="quicklinks_title">develop</td>
	   <td class=quicklinks><a href="http://sourceforge.net/projects/unison-db/">SourceForge</a> | <a href="http://sourceforge.net/project/showfiles.php?group_id=140591">download code &amp; data</a></td>
	  </tr>

	  <tr>
	   <td class="quicklinks_title">tools</td>
	   <td class=quicklinks><a href="javascript:addUnisonEngine()">Firefox search engine</a></td>
	  </tr>

	 </table>
	</td>
  </tr>
</table>
<!-- quicklinks end -->


<hr>

<h1>Unison Beyond the One-Liner</h1>


<h2>Contents</h2>

<ul>

<li>&gt;6M <i>non-redundant</i> (exact identity) sequences of varying
qualities from more than sources and thousands of species

<li>more than 6 CPU-years' worth of precomputed results for HMM alignments
(primarily Pfam), Prospect threading alignments<span style="color: red;">*</span>, TMHMM 2.0c transmembrane
detection, SignalP 3.0 signal prediction, cellular localization
(protcomp), genomic localization, regular expression motifs, and
others

<li>integrated patent info<span style="color: red;">*</span>

<li>indexes designed for rapid feature-based mining and hypothesis testing

<li>auxiliary data from SCOP, GO, HomoloGene, MINT, PDB to enable
richer interpretation of results and more expressive queries

<li>bookkeeping (run histories) and automated load/update procedures

<li>web front-end to facilitate sequence analysis and <i>limited</i>
feature-based mining

</ul>

<div style="padding: 5px; margin-left: 20px; margin-right: 20px; border: thin solid
red;"><span style="color: red;">*</span> The public version of Unison
contains only public sequences and results of non-proprietary methods.
The entire schema and loading tools are included with the public release;
institutions may load these proprietary data if they wish.</div>


<h2>Unison Goals</h2>

Unison was designed with these goals:

<ul>

<li>enable rapid, incremental updating of sequence analysis results

<li>use the sequence analysis for feature-based mining and to facilitate
hypothesis generation

<li>provide a free, multi-purpose platform for study of protein families
and phenomena and for the development of tools (e.g., our SNP-on-structure
viewer)

</ul>



<h2>The Build Process</h2>

<p>Sequences in Unison are culled from numerous sources and include both
well-annotated sequences (<i>e.g.,</i> from UniProt, Derwent Geneseq) and
speculative sequences (<i>e.g.,</i> from Ensembl, gene transcript
predictions, raw 6-frame translations).  The intent is to provide a
superset of all available sequences. Sequences are stored
<i>non-redundandtly</i> and are given a permanent identifier, a
<code>pseq_id</code>.  Sequences are never deleted or modified, and the
corresponding <code>pseq_id</code> is never changed. [<a
href="pseq_summary.pl?pseq_id=76">example</a>]

<p>Although sequences in Unison are stored non-redundantly, they came from
many highly redundant databases. Each of those databases typically
provides an accession number, a name, and a description for the sequence.
Unison stores this annotation with each sequence and the date that the
sequence was loaded from each source database and this provides a record
of when a sequence was loaded. Furthermore, each origin is ranked for its
reliability and informative content, and this is used to provide a
heuristic "best annotation" for each sequence as seen in the example.  [<a
href="pseq_paliases.pl?pseq_id=76">example</a>]

<p>After sequences are loaded, a set of reliable sequences is built based
on origin, species, and other properties.  These sequences are subjected
to an array of algorithms whose results are stored.  The predictions
currently include signal sequence, transmembrane prediction, GPI
anchoring, antigenicity, and alignments from BLAST, HMM (Pfam) alignments,
and threading, and many others.  Unison's bookkeeping about which
algorithms have been run (and how) provides a powerful mechanism for
keeping Unison sequences and results up to date [<a
href="pseq_history.pl?pseq_id=76">example</a>].

<p>All precomputed results are linked directly to the <code>pseq_id</code> [<a
href="pseq_features.pl?pseq_id=76">example</a>].  And because precomputed
results are linked to immutable sequences, there's no issue about results
being "stale" with respsect to a changed sequence.  Unison also
distinguishes results by algorighm parameters, and this permits users to
compare several sets of analyses.

<p>Unison integrates many other data sources which are useful in
querying. For example, NCBI's Homologene enables queries to require that a
human putative target also has a homolog in mouse or rat with similar
protein features.  The Structural Classification of Proteins (SCOP),
Patents, Gene Ontology, NCBI taxonomy, and genomic localization and
clustering are also available in Unison.





Unison contains 1) a comprehensive, non-redundent superset of protein
sequences from many sources, 2) numerous types of precomputed predictions
on those sequences, 3) auxiliary data that enables rich interpetations of
prediction results.

<p>The integration of these data enables mining of sequences based on
holistic protein feature criteria, synthesis of these predictions for the
analysis of individual sequences, and refinement of hypotheses regarding
the composition of protein families.


<p><b>Contact</b>
<p>Please <a
href="mailto:reece@harts.net,rkh@gene.com?subject=Unison">email Reece
Hart</a> with questions.


</body>
</html>