<html>
<head>
<title>Unison Protein Database</title>
<link rel="shortcut icon" href="favicon.png" />
<link rel="stylesheet" type="text/css" href="unison.css" />
</head>

<body>

<img src="logo5.png">
<br><font size="+2"><b>Unison Protein Database</b></font>


<h3>WHAT IS UNISON, IN A NUTSHELL?</h3>

Unison is a database of non-redundant protein sequences and extensive
<i>precomputed</i> predictions for those sequences.  Unison's primary
mission is to enable Genentech to ask and answer proteomic queries for
target discovery more efficiently, more completely, more reliably, and
more often, and to enable queries which were previously impractical.


<h3>WHAT CAN I DO WITH UNISON?</h3>
<dl>
  <dt><b>Answer complex, integrative queries efficiently.</b> <dd>Because
  precomputed results and auxiliary data are integrated within one
  resource, it is possible construct holistic queries which span multiple
  types and sources of data. For example: Identify distinct sequences
  which contain an extracellular immunoglobulin (ig) domain by Pfam,
  followed by a transmembrane region, followed by an Immunoreceptor
  Tyrosine Inhibitory Motif (ITIM).

  <dt><b>Analyze protein structure models.</b>
  <dd>Unison provides comprehensive sequence-sequence, sequence-profile,
  sequence-structure, and structure-structure alignments. These provide
  mechanisms to, for example, explore the structural effects of sequence
  variants (even on modeled structures) and to study the sequence
  divergence among structurally related proteins. <!--[SNP EXAMPLE]-->

  <dt><b>Generate <i>in silico</i> hypotheses.</b>
  <dd>Precomputed results also enable rapid responses to "what-if?"
  questions and the formulation of new hypotheses. For example: What
  Pfam domains occur upstream of a transmembrane domain and an ITIM?
  Unison answers this in 30 seconds.
</dl>


<h3>HOW DO I GET STARTED?</h3>

The Unison schema, source code, web pages, and non-proprietary data are
released under the Academic Free License.  You may download all of these
from the <a href="http://sourceforge.net/projects/unison-db/">Unison
SourceForge project pages</a>

<p>Tom Ferrin's group at UCSF has graciously agreed to host a
non-proprietary version Unison database. This site is being established
and I expect it to be available by August 1, 2005.


<h3>WHY?</h3>

Genentech undertook a large-scale effort to use structure prediction
methods to identify therapeutic targets in specific protein
families. Unison was borne of the need to manage and mine this large and
diverse set of data.

<p>Please contact Reece Hart &lt;<a
href="mailto:rkh@gene.com?Subject=Unison">rkh@gene.com</a>&gt; for more
information.


<h3>ACKNOWLEDGMENTS</h3>

Thanks to Genentech for providing a vibrant, challenging, and supportive
environment. Thanks also to Kiran Mukhyala and David Cavanaugh for their
many contributions to Unison, and to Bill Wood and the Bioinformatics
Department for support.

<hr>

<table width="100%" border=1><tr>
<td width="25%" align="center"><a href="http://www.gene.com"><img width=129 height=30 border=0 src="gene_tag_rgb.jpg" alt="Genentech"/></a></td>
<td width="25%" align="center">Bioinformatics Public Share<br><a href="http://share.gene.com">share.gene.com</a></td>
<td width="25%" align="center"><A href="http://sourceforge.net/"><IMG src="http://sourceforge.net/sflogo.php?group_id=4546&type=1" border="0" width="88" height="31"></A></td>
<td width="25%" align="center"><a href="http://www.postgresql.org"><img height="30" border=0 src="http://pgfoundry.org/images/elephantSmall.png"/><br>PostgreSQL</a></td>
</tr>
</table>

</body>

</html>